File: Assignment5/es5-impl/BinarySearchTree.js
var BaseBinaryTree = require("./BaseBinaryTree").BaseBinaryTree;

/**
 * Construct a binary search tree
 */
function BinarySearchTree() {
    BaseBinaryTree.call(this, undefined);
}

/* make inheritance */
BinarySearchTree.prototype = Object.create(BaseBinaryTree);

/**
 * Insert a value into BST
 * @param value the value to be inserted
 */
BinarySearchTree.prototype.insert = function (value) {
    var insertRecursively = (function (value, node) {
        if (!node || !node.value) {
            node = {
                value: value,
                left: undefined,
                right: undefined
            };
        }
        if (value == node.value)
            node.value = value;
        else if (value < node.value)
            node.left = insertRecursively(value, node.left);
        else node.right = insertRecursively(value, node.right);
        return node;
    }).bind(this);

    this.root = insertRecursively(value, this.root);
}

/**
 * Find a value in BST and get it
 * @param value The value to find
 * @returns the value to be found
 */
BinarySearchTree.prototype.get = function (value) {
    var getRecursively = (function (value, node) {
        if (!node)
            return null;
        if (value == node.value)
            return node.value;
        else if (value < node.value)
            return getRecursively(value, node.left);
        else return getRecursively(value, node.right);
    }).bind(this);

    return getRecursively(value, this.root);
}

/**
 * Remove a value in BST
 * @param value The value to be removed 
 * @returns the value removed
 */
BinarySearchTree.prototype.remove = function(value) {
    var removeRecursively = (function (value, node) {
        if (!node)
            return null;
        if (value == node.value) {
            if (!node.left)
                return node.right;
            else if (!node.right)
                return node.left;
            else {
                var rightNodeInLeftTree = node.left;
                while (rightNodeInLeftTree && rightNodeInLeftTree.right)
                    rightNodeInLeftTree = rightNodeInLeftTree.right;
                node.value = rightNodeInLeftTree.value;
                node.left = removeRecursively(value, node.left);
            }
        }
        else if (value < node.value)
            node.left = removeRecursively(value, node.left);
        else node.right = removeRecursively(value, node.right);
        return node
    }).bind(this);

    return removeRecursively(value, this.root);
}

/**
 * Iterate the BST
 * @param callback callback function
 */
BinarySearchTree.prototype.forEach = function (callback) {
    var stack = [], traversalNode = this.root;
    while (traversalNode || stack.length > 0) {
        while (traversalNode) {
            stack.push(traversalNode);
            traversalNode = traversalNode.left;
        }
        var node = stack.pop();
        traversalNode = node.right;
        callback(node.value);
    }
}

module.exports = {
    BinarySearchTree: BinarySearchTree
};

File: Assignment5/java-impl/BinarySearchTree.java
import java.util.Iterator;

/**
 * @author Huang Kaisheng (2020215138@stu.cqupt.edu.cn)
 * @version 1.0
 */
public class BinarySearchTree<T extends Comparable<T>> extends BaseBinaryTree<T> implements Iterable<T> {
    /**
     * Initialize the tree
     */
    public BinarySearchTree() {
        super();
    }

    /**
     * Insert a new value
     * 
     * @param value the value to be inserted
     */
    public void insert(T value) {
        this.root = insert(value, this.root);
    }

    private TreeNode<T> insert(T value, TreeNode<T> node) {
        if (node == null || node.value == null) {
            node = new TreeNode<T>(value);
            return node;
        }
        int compareToResult = value.compareTo(node.value);
        if (compareToResult == 0) {
            node.value = value;
        } else if (compareToResult < 0) {
            node.left = insert(value, node.left);
        } else {
            node.right = insert(value, node.right);
        }
        return node;
    }

    /**
     * Search the value
     * 
     * @param key the key to find the value
     * @return the wanted value
     */
    public T get(T key) {
        return get(key, this.root);
    }

    private T get(T key, TreeNode<T> node) {
        if (node == null) {
            return null;
        }
        int compareToResult = key.compareTo(node.value);
        if (compareToResult == 0) {
            return node.value;
        } else if (compareToResult < 0) {
            return get(key, node.left);
        } else {
            return get(key, node.right);
        }
    }

    /**
     * Remove a value from the tree
     * 
     * @param value the value to be removed
     */
    public void remove(T value) {
        root = remove(value, this.root);
    }

    private TreeNode<T> remove(T key, TreeNode<T> node) {
        if (node == null) {
            return null;
        }
        int compareToResult = key.compareTo(node.value);
        if (compareToResult == 0) {
            if (node.left == null) {
                return node.right;
            } else if (node.right == null) {
                return node.left;
            } else {
                TreeNode<T> rightNodeInLeftTree = node.left;
                while (rightNodeInLeftTree != null && rightNodeInLeftTree.right != null)
                    rightNodeInLeftTree = rightNodeInLeftTree.right;
                node.value = rightNodeInLeftTree.value;
                node.left = remove(key, node.left);
            }
        } else if (compareToResult < 0) {
            node.left = remove(key, node.left);
        } else {
            node.right = remove(key, node.right);
        }
        return node;
    }

    /**
     * Construct an iterator of this tree
     * 
     * @return an iterator to in-order traverse this tree
     */
    public Iterator<T> iterator() {
        return new TreeIterator<T>(this);
    }

    /**
     * Return string representation of this tree
     * 
     * @return a string that represent this tree
     */
    public String toString() {
        return "root: \n" + this.root.toString();
    }
}
